### ISSUE TITLE
Pickling AutoReject objects and parallelizing with joblib fails

### ISSUE THREAD

**Issue by hubertjb** [autoreject/autoreject#192]
I was trying to parallelize the application of an `AutoReject` object using joblib. I noticed both `Parallel` and pickling functions fail, due to missing attributes. Here's a snippet to reproduce the error:

```python
import mne
from joblib import Parallel, delayed, dump, load
import numpy as np
from autoreject import AutoReject

# Create dummy Epochs object
info = mne.create_info(ch_names=['F1', 'C1', 'Cz', 'C2', 'F2'], sfreq=100,
                       ch_types=['eeg'] * 5)
epochs_data = np.random.rand(10, 5, 100)
epochs = mne.EpochsArray(epochs_data, info=info)
epochs.set_montage('standard_1020')

ar = AutoReject()

# This fails:
Parallel(n_jobs=2)(delayed(ar.fit_transform)(ep) for ep in [epochs, epochs])

# This fails as well:
fname = './test_ar.pkl'
dump(ar, fname)
loaded_ar = load(fname)
print(loaded_ar)
```

The traceback I get with this snippet:

```
The above exception was the direct cause of the following exception:

AttributeError                            Traceback (most recent call last)
~/Documents/ssleep/dev_tests/test_autoreject_joblib.py in <module>
     18 
     19 # This fails:
---> 20 Parallel(n_jobs=2)(delayed(ar.fit_transform)(ep) for ep in [epochs, epochs])
     21 
     22 # This fails as well:

~/miniconda3/envs/braindecode/lib/python3.8/site-packages/joblib/parallel.py in __call__(self, iterable)
   1040 
   1041             with self._backend.retrieval_context():
-> 1042                 self.retrieve()
   1043             # Make sure that we get a last message telling us we are done
   1044             elapsed_time = time.time() - self._start_time

~/miniconda3/envs/braindecode/lib/python3.8/site-packages/joblib/parallel.py in retrieve(self)
    919             try:
    920                 if getattr(self._backend, 'supports_timeout', False):
--> 921                     self._output.extend(job.get(timeout=self.timeout))
    922                 else:
    923                     self._output.extend(job.get())

~/miniconda3/envs/braindecode/lib/python3.8/site-packages/joblib/_parallel_backends.py in wrap_future_result(future, timeout)
    540         AsyncResults.get from multiprocessing."""
    541         try:
--> 542             return future.result(timeout=timeout)
    543         except CfTimeoutError as e:
    544             raise TimeoutError from e

~/miniconda3/envs/braindecode/lib/python3.8/concurrent/futures/_base.py in result(self, timeout)
    437                 raise CancelledError()
    438             elif self._state == FINISHED:
--> 439                 return self.__get_result()
    440             else:
    441                 raise TimeoutError()

~/miniconda3/envs/braindecode/lib/python3.8/concurrent/futures/_base.py in __get_result(self)
    386     def __get_result(self):
    387         if self._exception:
--> 388             raise self._exception
    389         else:
    390             return self._result

AttributeError: 'AutoReject' object has no attribute 'picks'
```

I've narrowed it down to this line in the definition of `AutoReject`: https://github.com/autoreject/autoreject/blob/master/autoreject/autoreject.py#L935
It looks like `__setstate__` ignores attributes that are in the `_INIT_PARAMS` dictionary, and so after the unpickling process these attributes are missing from the object.

@agramfort

**Comment by jasmainak** [comment]
If you can propose a fix with a test, that would be more than welcome!

**Comment by hubertjb** [comment]
Hi @jasmainak, perfect, I'll take a shot at it. A quick question before: I realized the if condition [here](https://github.com/autoreject/autoreject/blob/master/autoreject/autoreject.py#L935) is there so that when using the `save()` method followed by `read_auto_reject` the object can first be instantiated with the right init attributes, and then its state can be set with `__setstate__`. Would it make sense to instead instantiate the object with default parameters, and then set all attributes at once (including those of the init method) with one call to `__setstate__`? Allowing that would fix the pickling issue as well.

**Comment by jasmainak** [comment]
Give it a try. It sounds like a reasonable approach. If the tests fail, we'll know something is wrong :)

**Comment by agramfort** [comment]
I am also curious why we need to do this in the first place.

>
can you also try to remove the __setstate__ etc. methods to see where it
fails?

**Comment by hubertjb** [comment]
It looks like we need `__setstate__` and `__getstate__` methods (or something similar) for two reasons: (1) they are necessary if we want the `save` method to write to a dictionary in an h5 file, and (2) they take care of the instantiation of `_AutoReject` objects for each channel type in a nested dictionary structure.

Concerning (1), I guess `__getstate__` and `__setstate__` could be replaced by `to_dict` and `from_dict` methods that work essentially in the same way. This would avoid potential collisions with what pickle needs. However, this clashing can currently be fixed with a one-line change in `__setstate__`, so I'm not sure such a change would be necessary I don't see a way to avoid (2) though, unless the io functionality was implemented with pickle in the first place - in that case we could drop the `__setstate__` and `__getstate__` methods entirely.

**Comment by jasmainak** [comment]
I think the idea was that you could save with pickle or hdf5 as you like and they would give you identical results. Not sure if we ever tested that though

**Comment by agramfort** [comment]
Whatever works

**Comment by jasmainak** [comment]
closed by #193
